### PR TITLE
fix58438: do not delete comments/code following unterminated string completion

### LINKED ISSUE #58438: Selecting suggestion of string completion deletes everything up to end of line, if no ending quote is present
### 🔎 Search Terms

completion autocompletion string

### 🕗 Version & Regression Information

- This issue is new in typescript 5.5.0-beta
- Not fixed in nightly at time of submitting issue.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.5.0-beta#code/C4TwDgpgBAKhDOwoF4oHJgOGg3AKAHoCoAFAGwgEN4IAaKYAJxAYHsoBjVgWzAswYALaJkRQARiwDCTMgFp4YSh2iUAdgBMhENVBoUOwAJZqA5noCup01iOs1eLmrGjgALlhYU6bDihEYQSN4KAB3VkYAaxCAMxMIR3sXLAAmDzgxVAw-IlggkI0IOLUjTDIWeEFWCzItNVYkGyRC-ggte04ePghje3pxCA5KCxooUs6arQGx3gjgdSQuQv6LJHGNVgRE5yRXAGZ0r1QAImAc4kDgqELi0ohyvSrJqHrGnuv7nraoDq5efjsan6g2Go3GXGe0yMs0Y8zUi1YywkqzGzU28DwhGIACUispgBETOYIAAPSj-aChYSMET5CQQQSUABudkYYxC4koGgAdLBmHpPoYiZZrLYOjFGDwoKdjmxpa5jttkogACweADeUDAkrAAEYPGoLNwBox6NrWGA0i8jSaoABfbzq816jwAVjNOqtKXt+CcyuAro1zoADIdEB6LfrrcaIKatZ6DTbY-R7Y74xbQz4oLRnVH3enLR5vXb8LkpF0AR0TFBXdzXVAafAasAQgSoErdlgAGxBnWZjLACMu6MmodWw0xuOp1CakMeDBYLFQcsU3q6au1lV1hsIZut9gdhhYADsvYzYcHBajE9HBfHSanDpnBczC7EOZ1ebHRZ9QA

### 💻 Code

```ts
type Test = 'test';
// Please, try to complete the test by Ctrl-space and then selecting suggestion
const test: Test = 't'; //This works fine
const test2: Test = 't; // This comment definitely should not get deleted on completion, because it could be important code, but it does
const test3: Test = "t; // This comment  definitely should not get deleted on completion, because it could be important code, but it does

// Refactoring example where this behavior is bad. Try selecting suggestion from "t" to "test"
const test4: { prop1: number, prop2: number } = {prop1: 5, prop2: 2 };
const test5: {prop0: Test, prop1: number, prop2: number,  } = { prop0: 't ,prop1: 5, prop2: 2 };
// Completion in 5.5 results to 
const test6: {prop0: Test, prop1: number, prop2: number,  } = { prop0: 'test
// Completion in 5.4.5 results to 
const test7: {prop0: Test, prop1: number, prop2: number,  } = { prop0: 'test ,prop1: 5, prop2: 2 };
```

### 🙁 Actual behavior

When trying to complete "t" to "test" in the code for variables test2, test3, everything up to the end of line gets deleted.

This is especially bad when trying to add new string properties to existing objects, as can be seen by example refactoring of test4 to test5 variable. Where poor result of selected suggestion can be seen in test6 variable and good behavior of previous TS versions in test7 variable.

### 🙂 Expected behavior

When editing new string, typescript should insert suggestions, not replace content. Previous versions behavior seems perfect to me.

### Additional information about the issue

EDIT: updated example, one of the previous examples didn't behave very well even in previous versions. Added demonstration of previous good behavior.

_No response_

### PR DESCRIPTION
**Pull request by iisaduan**
fixes #58438

The above issue is a regression bug that happened because of #57839, which was included with 5.5 beta. This PR updates the `replacementRange` to be more inline with 5.4 behavior, namely completions on unterminated strings will not remove anything to the right of the cursor. Closed strings will replace everything within the quotes.

Examples: 

```typescript
type Test = `abcde';
const test: Test = 'a|; //comment
const test2: Test = 'a|bc // comment
const test3: Test = 'a|bc' //comment
```
becomes
```typescript
type Test = `abcde';
const test: Test = 'abcde; //comment
const test2: Test = 'abcdebc // comment
const test:3 Test = 'abcde'; //comment
```
after completions are triggered at all three locations.

### CODE REVIEW

File: tests/cases/fourslash/stringCompletionsUnterminated2.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+///<reference path="fourslash.ts" />
+
+
+
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/stringCompletionsUnterminated2.ts:4]:
```suggestion
```

File: tests/cases/fourslash/stringCompletionsUnterminated2.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// @Filename: test7.ts
+//// type Test = 'test asdf';
+//// const test7: Test = `[|test as/*7*/|]; // this comment should not be deleted
+
+// @Filename: test8.ts
+//// type Test = 'test asdf';
+//// const test8: Test = `[|test/*8*/|] as; // this comment should not be deleted
+
+
+
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/stringCompletionsUnterminated2.ts:42]:
```suggestion
```

File: src/services/utilities.ts [hunk 3]
```diff
@@ -2335,12 +2335,12 @@ export function createTextSpanFromNode(node: Node, sourceFile?: SourceFile, endN
...
 
 /** @internal */
-export function createTextSpanFromStringLiteralLikeContent(node: StringLiteralLike) {
+export function createTextSpanFromStringLiteralLikeContent(node: StringLiteralLike, position?: number) {
     let replacementEnd = node.getEnd() - 1;
     if (node.isUnterminated) {
         // we return no replacement range only if unterminated string is empty
         if (node.getStart() === replacementEnd) return undefined;
-        replacementEnd = node.getEnd();
+        replacementEnd = position !== undefined && position <= node.getEnd() ? position : node.getEnd();
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts:2343]: There are only two callers of `createTextSpanFromStringLiteralLikeContent`, and neither of them omit the `position` argument, so just make it required (and document what `position` is supposed to mean)

Comment (DanielRosenwasser) [reply src/services/utilities.ts:2343]: Also, do you have any tests for when `position` is strictly greater than the end of the node? I don't think I'm seeing any.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: src/services/utilities.ts (hunk 3 above)
Comment (iisaduan) [reply src/services/utilities.ts:2343]: Ah, that was old code, thanks for catching it! (I tried to make `position` optional, but it ended up being needed in more places)

Comment (iisaduan) [reply src/services/utilities.ts:2343]: It shouldn't be possible _in a completion_ for `position > node end` but I will add that check incase `createTextSpanFromStringLiteralLikeContent` will be used in a non-completion context in the future.

Comment (DanielRosenwasser) [reply src/services/utilities.ts:2343]: The only thing I can think of is possibly an unterminated string literal on the previous line

```ts
let s: "string" = "s
/*$*/
```

Comment (iisaduan) [reply src/services/utilities.ts]: Completions won't be triggered on new lines -- so 
```typescript
type LongString = `asdf
string
line`

let s: LongString = `asdf
/**/
  ^ no suggestions

let r: LongString = `asdf/**/
                          ^ completes to `asdf\nstring\nline`
```

File: src/services/utilities.ts [hunk 4]
```diff
@@ -2335,12 +2335,12 @@ export function createTextSpanFromNode(node: Node, sourceFile?: SourceFile, endN
...
 
 /** @internal */
-export function createTextSpanFromStringLiteralLikeContent(node: StringLiteralLike) {
+export function createTextSpanFromStringLiteralLikeContent(node: StringLiteralLike, position: number) {
     let replacementEnd = node.getEnd() - 1;
     if (node.isUnterminated) {
         // we return no replacement range only if unterminated string is empty
         if (node.getStart() === replacementEnd) return undefined;
-        replacementEnd = node.getEnd();
+        replacementEnd = position <= node.getEnd() ? position : node.getEnd();
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts:2343]:
```suggestion
        replacementEnd = Math.min(position, node.getEnd());

```

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick to release-5.5

**Pull request merged**